### PR TITLE
Support Path Attribute in SCIM2 Patch Add Operation

### PR DESCRIPTION
**Pull request by sarubi**
## Purpose
Fix: wso2/product-is#5321

There are three operations  "add", "remove", or "replace" can be done via PATCH in SCIM.  
The "path" attribute is OPTIONAL for "add" and "replace" and is REQUIRED for "remove" operations. 
We already support the "path" attribute for PATCH Replace and Remove. In this PR we support it for PATCH add as well. 
The "path" attribute value is a String containing an attribute path describing the target of the operation.  
Path attribute can contain a path to an attribute or can have filters. Please find the valid path examples as follows:
```
       "path":"members"

       "path":"name.familyName"

       "path":"addresses[type eq \"work\"]"

       "path":"members[value eq
              \"2819c223-7f76-453a-919d-413861904646\"]"

       "path":"members[value eq
              \"2819c223-7f76-453a-919d-413861904646\"].displayName"

```

1. We supported one to three consecutive path attribute specified (without filters). 
Ex:  "Path" : "attributeX"
       "Path" : "attributeX.subAttributeY"
2. When the filter is specified in the path, then we handle those cases as PATCH REPLACE with filters which are already implemented here.
Ex:  "path":"AttributeX[AttributeXX eq \"AttributeXXExpectedValue\"]"

Also **Integrated an ABNF parser to validate the provided path syntax** before performing PATCH Add Operation. Used [parse2](http://www.parse2.com/) to generate parser classes according to the rules specified in the SCIM specification, these parser classes located under the src/gen with the package name `org.wso2.charon3.core.aParser`.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by sarubi** [comment]
Due to the major refactor in the PatchOperationUtil.java changes comparison is going wrong here, please find the code snippet that was newly added for the patch add operation below,

https://github.com/sarubi/charon/blob/439d089c6cb8db3ce4540e0df7f47abdf8b5b259/modules/charon-core/src/main/java/org/wso2/charon3/core/utils/PatchOperationUtil.java#L645-L1381

**Comment by sarubi** [comment]
Please find the code snippet that was newly added for the patch add operation below,

https://github.com/wso2/charon/blob/7a7c05fc1518e823009b8f1f5e618f559bc86f87/modules/charon-core/src/main/java/org/wso2/charon3/core/utils/PatchOperationUtil.java#L646-L1531

File: modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/package-info.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * under the License.
+ */
+
+/**
+ * This package classes generated using aparse-2.5.jar which is a free parser tool called
+ * parse2( http://www.parse2.com/).
+ * This parser parses the path attribute according to the ABNF rules specified in the SCIM Specification.
+ * Rules used to generate this PATH parser can be found in path-abnf-rules.abnf file.
+ */
+package org.wso2.charon3.core.aParser;
```
Comment (IsuraD) [inline modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/package-info.java]: new line

File: modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/path-abnf-rules.abnf [hunk 2]
```diff
@@ -0,0 +1,69 @@
+PATH = attributePath / valuePath [subAttribute];
```
Comment (IsuraD) [inline modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/path-abnf-rules.abnf]: put license header

File: modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/package-info.java (hunk 1 above)
Comment (sarubi) [reply modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/package-info.java]: Fixed

File: modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/path-abnf-rules.abnf (hunk 2 above)
Comment (sarubi) [reply modules/charon-core/src/gen/java/org/wso2/charon3/core/aParser/path-abnf-rules.abnf]: added

**Pull request merged**